README.md
# haskell-dev-sprint-calculator


CHANGELOG.md
# Changelog for `haskell-dev-sprint-calculator`

All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/),
and this project adheres to the
[Haskell Package Versioning Policy](https://pvp.haskell.org/).

## Unreleased

## 0.1.0.0 - YYYY-MM-DD


Config.hs
{-# LANGUAGE OverloadedStrings #-}
{-# LANGUAGE QuasiQuotes #-}

module Config where

-- https://ro-che.info/articles/2015-07-26-better-yaml-parsing
-- for [r|raw string goes here|] ext to QuasiQuotes

import Control.Applicative
import Data.ByteString (ByteString)
import qualified Data.ByteString.Char8 as BS
import Data.Map
import Data.Maybe (fromJust)
import Data.Yaml
import qualified Data.Yaml as Y
import Text.RawString.QQ

-- See https://github.com/deepin-community/haskell-yaml

configYaml :: ByteString
configYaml =
  [r|
baseline:
  date: 2023-01-01
  sprint: 123
|]

data SprintBase = SprintBase
  { date :: String,
    sprint :: Int
  }
  deriving (Show, Eq)

newtype Baseline = Baseline {baseline :: SprintBase}

-- data Baseline = Baseline
--  { baseline :: SprintBase
--  }
--  deriving (Eq, Show)

-- instance FromJSON Baseline where
--  parseJSON = withObject "Baseline" $ o -> Baseline <$>
--    o .: "baseline"
--    <*> o .: "age"

-- Could use Generic deriving, but doing by hand here:
instance FromJSON Baseline where
  parseJSON (Object o) = Baseline <$> (o .: "baseline")
  parseJSON _ = error "Can't parse Task from YAML/JSON"

instance ToJSON Baseline where
  toJSON (Baseline c) = "baseline" .= c

type ErrorMsg = String

readMyType :: ByteString -> Either ErrorMsg Baseline
readMyType yamlObj = do
  parsedContent <- decodeEither yamlObj

  -- mapOfStrings <- parseEither (.: "baseline") yamlObj
  case parsedContent of
    Right yamlObj' -> Right yamlObj'
    Left exception -> Left $ show (exception)

-- instance FromJSON Baseline where
--   parseJSON (Y.Object v) = -- called parseJSON because it derives from the aeson lib
--     Baseline <$>
--       v .: "baseline"
--       <*> Baseline <$>
--         v .: "date" <*>
--         v .: "sprint"
--   parseJSON _ = error "Cannot parse 'baseline' from YAML/JSON"
--
-- Cool wrapper I found https://cthulehansen.github.io/blog/posts/jsonyaml.html
yamlDecode :: (FromJSON a) => ByteString -> Either String a
yamlDecode a = case Data.Yaml.decodeEither' a of
  Left pe -> Left $ show pe
  Right a' -> Right a'

main = do
  -- configYaml <- BS.readFile "config.yml"
  -- let config = Data.Yaml.decode configYaml :: Maybe [Baseline]
  -- config <- Data.Yaml.decodeThrow configYaml
  -- parsedConfig <- readMyType (configYaml)
  case yamlDecode configYaml of
    Right d -> putStrLn d
    Left e -> putStrLn e


examples/Config_Generics_Deriving_JSON_DTOs.hs
{-# LANGUAGE OverloadedStrings #-}
{-# LANGUAGE QuasiQuotes #-}
-- Generic is a typeclass that allows GHC to represent your types at a level of generic constructors.
-- In other words, we do not need to manually define the FromJSON objects like we
-- did above if we add derive (Generic, ToJson, FromJSON) to the data records.
-- {-# LANGUAGE DeriveGeneric     #-}
-- {-# LANGUAGE DeriveAnyClass    #-}
{-# LANGUAGE DeriveGeneric     #-}
{-# LANGUAGE DeriveAnyClass    #-}

module Config_Generics_Deriving_JSON_DTOs where

-- https://mmhaskell.com/blog/2017/6/5/flexible-data-with-aeson
-- https://ro-che.info/articles/2015-07-26-better-yaml-parsing
-- for [r|raw string goes here|] ext to QuasiQuotes
import GHC.Generics

import Control.Applicative
import Data.ByteString (ByteString)
import qualified Data.ByteString.Char8 as BS
import Data.Map
import Data.Maybe (fromJust)
import Data.Yaml
import qualified Data.Yaml as Y
import Text.RawString.QQ
import Data.Time

-- See https://github.com/deepin-community/haskell-yaml

configYaml :: ByteString
configYaml =
  [r|
  base:
    date: 2023-01-01
    sprint: 123
  |]

-- Note this haskell representation of data can be used to 
-- auto gen FromJSON / ToJSON types at compile time.
data Config = Config {
  base :: StartingSprint
} deriving (Show, Generic, ToJSON, FromJSON)

data StartingSprint = StartingSprint {
  date   :: String, -- Day
  sprint :: Int
} deriving (Show, Generic, ToJSON, FromJSON)

-- baseHaskellRepresentationExample :: Config
-- baseHaskellRepresentationExample = Config
--   {
--     base = StartingSprint
--       {
--         date = "2023-01-01",
--         sprint = 123
--       }
--   }


-- {
--   "baseline" : {
--     "date": "2023-01-01"
--     "sprint": 123
--   }
-- }

-- aeson lib encapsulates all JSON types in
-- data Value =
--   Object (HashMap Text Value) |
--   Array (Vector Value) |
--   String Text |
--   Number Scientific |
--   Bool Bool |
--   Null


--
extractConfig :: IO Config
extractConfig = do
  case Y.decodeEither' (configYaml) of
    Right base -> return base
    Left err -> error ("Failure while trying to parse input: " ++ show err)

main :: IO ()
--     pseudo-java code using fpinjava:
--     extractConfig.flatMap(b -> convertToStringAndPrint)
main = extractConfig >>= \b -> putStrLn $ (show b) ++ "     <-- Config_Generics_Deriving_JSON_DTOs"



examples.disabled/Config_Haskell_Templates_Generate_JSON_DTOs.hs
{-# LANGUAGE OverloadedStrings #-}
{-# LANGUAGE QuasiQuotes #-}
-- read https://markkarpov.com/tutorial/th.html
{-# LANGUAGE TemplateHaskell #-}

module Config_Haskell_Templates_Generate_JSON_DTOs where

-- https://mmhaskell.com/blog/2017/6/5/flexible-data-with-aeson
-- https://ro-che.info/articles/2015-07-26-better-yaml-parsing
-- for [r|raw string goes here|] ext to QuasiQuotes

import Data.Aeson.TH (deriveJSON, defaultOptions) -- aeson >=2 incompatibility issues with Data.Yaml



import Control.Applicative
import Data.ByteString (ByteString)
import qualified Data.ByteString.Char8 as BS
import Data.Map
import Data.Maybe (fromJust)
import Data.Yaml
import qualified Data.Yaml as Y
import Text.RawString.QQ
import Data.Time

-- See https://github.com/deepin-community/haskell-yaml

configYaml :: ByteString
configYaml =
  [r|
  base:
    date: 2023-01-01
    sprint: 123
  |]

-- Note this haskell representation of data can be used to 
-- auto gen FromJSON / ToJSON types at compile time.
data Base = Base {
  base :: StartingSprint
} deriving (Show)

data StartingSprint = StartingSprint {
  date   :: String, -- Day
  sprint :: Int
} deriving (Show)

-- The two apostrophes before a type name is template haskell syntax
-- defaultOptions can be replaced with a record for more control
deriveJSON defaultOptions ''Base
deriveJSON defaultOptions ''StartingSprint

-- baseHaskellRepresentationExample :: Base
-- baseHaskellRepresentationExample = Base
--   {
--     base = StartingSprint
--       {
--         date = "2023-01-01",
--         sprint = 123
--       }
--   }


-- {
--   "baseline" : {
--     "date": "2023-01-01"
--     "sprint": 123
--   }
-- }

-- aeson lib encapsulates all JSON types in
-- data Value =
--   Object (HashMap Text Value) |
--   Array (Vector Value) |
--   String Text |
--   Number Scientific |
--   Bool Bool |
--   Null


--
extractBase :: IO Base
extractBase = do
  case Y.decodeEither' (configYaml) of
    Right base -> return base
    Left err -> error ("Failure while trying to parse input: " ++ show err)

main :: IO ()
--     pseudo-java code using fpinjava:
--     extractBase.flatMap(b -> convertToStringAndPrint)
main = extractBase >>= \b -> putStrLn $ (show b) ++ "     <-- Config_Haskell_Templates_Generate_JSON_DTOs"



examples/Config_With_Manual_JSON_DTOs.hs
{-# LANGUAGE OverloadedStrings #-}
{-# LANGUAGE QuasiQuotes #-}

module Config_With_Manual_JSON_DTOs where

-- https://mmhaskell.com/blog/2017/6/5/flexible-data-with-aeson
-- https://ro-che.info/articles/2015-07-26-better-yaml-parsing
-- for [r|raw string goes here|] ext to QuasiQuotes

import Control.Applicative
import Data.ByteString (ByteString)
import qualified Data.ByteString.Char8 as BS
import Data.Map
import Data.Maybe (fromJust)
import Data.Yaml
import qualified Data.Yaml as Y
import Text.RawString.QQ
import Data.Time

-- See https://github.com/deepin-community/haskell-yaml
-- https://abhinavsarkar.net/posts/json-parsing-from-scratch-in-haskell/#fn6

configYaml :: ByteString
configYaml =
  [r|
  base:
    date: 2023-01-01
    sprint: 123
  |]

-- Note this haskell representation of data can be used to 
-- auto gen FromJSON / ToJSON types at compile time.
-- using either Template Haskell (deriveJSON defaultOptions ''Config) or 
-- deriving Generic after enabling compile opts DeriveGeneric, DeriveAnyClass
data Config = Config {
  base :: StartingSprint
} deriving (Show)

data StartingSprint = StartingSprint {
  date   :: String, -- Day
  sprint :: Int
} deriving (Show)

-- baseHaskellRepresentationExample :: Config
-- baseHaskellRepresentationExample = Config
--   {
--     base = StartingSprint
--       {
--         date = "2023-01-01",
--         sprint = 123
--       }
--   }


-- {
--   "baseline" : {
--     "date": "2023-01-01"
--     "sprint": 123
--   }
-- }

-- aeson lib encapsulates all JSON types in
-- data Value =
--   Object (HashMap Text Value) |
--   Array (Vector Value) |
--   String Text |
--   Number Scientific |
--   Bool Bool |
--   Null

instance FromJSON Config where
  -- The parseJSON functions are defined through eta reduction. This is why
  -- there seems to be no parameter. The .: operator can grab any data type
  -- that conforms to FromJSON itself.
  parseJSON = withObject "Config" $ \o -> do
    base_ <- o .: "base"
    return $ Config base_

instance FromJSON StartingSprint where
  parseJSON = withObject "StartingSprint" $ \o -> do
    date_ <- o .: "date"
    sprint_ <- o .: "sprint"
    return $ StartingSprint date_ sprint_


-- Generic is a class that allows GHC to represent your types at a level of generic constructors.
-- In other words, we do not need to manually define the FromJSON objects like we
-- did above if we add derive (Generic, ToJson, FromJSON) to the data records.
-- {-# LANGUAGE DeriveGeneric     #-}
-- {-# LANGUAGE DeriveAnyClass    #-}
--
extractConfig :: IO Config
extractConfig = do
  case Y.decodeEither' (configYaml) of
    Right base -> return base
    Left err -> error ("Failure while trying to parse input: " ++ show err)

main :: IO ()
--     pseudo-java code using fpinjava:
--     extractConfig.flatMap(b -> convertToStringAndPrint)
-- main = extractConfig >>= print
main = extractConfig >>= \b -> putStrLn $ (show b) ++ "     <-- Config_With_Manual_JSON_DTOs"
